### PR TITLE
Implement no-yank delete/change

### PR DESCRIPTION
**Pull request by ath3**

### CODE REVIEW

File: book/src/keymap.md [hunk 1]
```diff
@@ -45,40 +45,42 @@
...
+| `p`         | Paste after selection                                      | `paste_after`             |
+| `P`         | Paste before selection                                     | `paste_before`            |
+| `"` `<reg>` | Select a register to yank to or paste from                 | `select_register`         |
+| `>`         | Indent selection                                           | `indent`                  |
+| `<`         | Unindent selection                                         | `unindent`                |
+| `=`         | Format selection (**LSP**)                                 | `format_selections`       |
+| `d`         | Delete selection                                           | `delete_selection`        |
+| `Alt-d`     | Delete selection (dont yank)                               | `delete_selection_noyank` |
+| `c`         | Change selection (delete and enter insert mode)            | `change_selection`        |
+| `Alt-c`     | Change selection (delete and enter insert mode, dont yank) | `change_selection_noyank` |
```
Comment (Omnikar) [inline book/src/keymap.md:74]:
```suggestion
| Key         | Description                                                 | Command                   |
| -----       | -----------                                                 | -------                   |
| `r`         | Replace with a character                                    | `replace`                 |
| `R`         | Replace with yanked text                                    | `replace_with_yanked`     |
| `~`         | Switch case of the selected text                            | `switch_case`             |
| `` ` ``     | Set the selected text to lower case                         | `switch_to_lowercase`     |
| `` Alt-` `` | Set the selected text to upper case                         | `switch_to_uppercase`     |
| `i`         | Insert before selection                                     | `insert_mode`             |
| `a`         | Insert after selection (append)                             | `append_mode`             |
| `I`         | Insert at the start of the line                             | `prepend_to_line`         |
| `A`         | Insert at the end of the line                               | `append_to_line`          |
| `o`         | Open new line below selection                               | `open_below`              |
| `O`         | Open new line above selection                               | `open_above`              |
| `.`         | Repeat last change                                          | N/A                       |
| `u`         | Undo change                                                 | `undo`                    |
| `U`         | Redo change                                                 | `redo`                    |
| `y`         | Yank selection                                              | `yank`                    |
| `p`         | Paste after selection                                       | `paste_after`             |
| `P`         | Paste before selection                                      | `paste_before`            |
| `"` `<reg>` | Select a register to yank to or paste from                  | `select_register`         |
| `>`         | Indent selection                                            | `indent`                  |
| `<`         | Unindent selection                                          | `unindent`                |
| `=`         | Format selection (**LSP**)                                  | `format_selections`       |
| `d`         | Delete selection                                            | `delete_selection`        |
| `Alt-d`     | Delete selection (don't yank)                               | `delete_selection_noyank` |
| `c`         | Change selection (delete and enter insert mode)             | `change_selection`        |
| `Alt-c`     | Change selection (delete and enter insert mode, don't yank) | `change_selection_noyank` |
```

Comment (Omnikar) [reply book/src/keymap.md:74]: (This just adds apostrophes and makes the whole table 1 column wider to accommodate).

File: helix-term/src/commands.rs [hunk 2]
```diff
@@ -227,7 +226,9 @@ impl Command {
         extend_line, "Select current line, if already selected, extend to next line",
         extend_to_line_bounds, "Extend selection to line bounds (line-wise selection)",
         delete_selection, "Delete selection",
+        delete_selection_noyank, "Delete selection (dont yank)",
         change_selection, "Change selection (delete and enter insert mode)",
+        change_selection_noyank, "Change selection (delete and enter insert mode, dont yank)",
```
Comment (Omnikar) [inline helix-term/src/commands.rs:231]:
```suggestion
        delete_selection_noyank, "Delete selection (don't yank)",
        change_selection, "Change selection (delete and enter insert mode)",
        change_selection_noyank, "Change selection (delete and enter insert mode, don't yank)",
```

**Comment by Omnikar** [comment]
Couldn't this have just been done by changing the `reg` parameter of `delete_selection_impl` to take an `Option<&mut Register>` and then adjusting the function accordingly?

**Comment by ath3** [comment]
Yes, and thats how i started doing it, but it felt like change and delete share a lot of the same code handling register, so i moved it to one central place (delete_selection_impl).
The only thing that differs in delete/change is what happens after the register related stuff and the deletion of text is done, so this feels more logical to me.

Btw the test is failing because sometimes git clone related errors occur, like it happened now.

File: helix-term/src/commands.rs [hunk 3]
```diff
@@ -1528,41 +1529,58 @@ fn extend_to_line_bounds(cx: &mut Context) {
...
+    let (view, doc) = current!(cx.editor);
+
     let text = doc.text().slice(..);
-    let selection = doc.selection(view_id);
+    let selection = doc.selection(view.id);
 
-    // first yank the selection
-    let values: Vec<String> = selection.fragments(text).map(Cow::into_owned).collect();
-    reg.write(values);
+    if with_reg {
```
Comment (archseer) [inline helix-term/src/commands.rs:1543]: Let's make a "black hole" register `'_'`. Then the extra parameter can be removed, and this condition can be `if cx.register == Some('_')`. You can then explicitly set the register in the `noyank` versions

File: helix-term/src/commands.rs [hunk 4]
```diff
@@ -227,7 +226,9 @@ impl Command {
         extend_line, "Select current line, if already selected, extend to next line",
         extend_to_line_bounds, "Extend selection to line bounds (line-wise selection)",
         delete_selection, "Delete selection",
+        delete_selection_noyank, "Delete selection (don't yank)",
         change_selection, "Change selection (delete and enter insert mode)",
+        change_selection_noyank, "Change selection (delete and enter insert mode, don't yank)",
```
Comment (archseer) [inline helix-term/src/commands.rs:231]: Let's instead use ", without yanking" for both of these.

https://github.com/mawww/kakoune/blob/master/src/normal.cc#L2236-L2239

**Review by archseer — APPROVED** [review]
🎉

**Comment by archseer** [comment]
Ah, looks like there's a conflict now, can you resolve it? Good to merge after that

**Comment by ath3** [comment]
Done!

**Comment by sudormrfbin** [comment]
(The black hole register needs to be added to https://docs.helix-editor.com/usage.html#special-registers)

**Comment by ath3** [comment]
Black hole register is a special case for noyank versions of change and delete, so outside this they are still usable.
Should i change them to behave globally as a black hole?

**Comment by ath3** [comment]
Maybe that would be a good idea, because this might be confusing for users, a register behaving differently for only some functions.

**Comment by sudormrfbin** [comment]
Yeah giving `_` the role of the black hole register globally and explicitly would be good.

**Comment by pickfire** [comment]
I wonder why not `"_d`?

**Pull request merged**